### ISSUE TITLE
AWS Cognito app client resource unchecks options that have been enabled on alternative applies

### ISSUE THREAD

**Issue by karthikkp** [terraform-providers/terraform-provider-aws#5029]
**Community Note**

- Please vote on this issue by adding a 👍 reaction to the original issue to help the community and maintainers prioritize this request

- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

**0.10.8**

### Affected Resource(s)

* aws_cognito_user_pool_client

### Terraform Configuration Files

```
resource "aws_cognito_user_pool_client" "client" {
  name         = "name"
  user_pool_id = "${aws_cognito_user_pool.pool.id}"

  generate_secret                      = true
  explicit_auth_flows                  = ["ADMIN_NO_SRP_AUTH", "USER_PASSWORD_AUTH"]
  allowed_oauth_flows_user_pool_client = true
  allowed_oauth_flows                  = ["code"]
  supported_identity_providers         = ["COGNITO"]
  allowed_oauth_scopes                 = ["aws.cognito.signin.user.admin", "openid", "profile", "email"]

  callback_urls = ["https://example.com"]
}
```

### Expected Behavior

1. The resource should not get modified if there are no changes
2. If there are no changes to these fields "allowed_oauth_flows, allowed_oauth_scopes, supported_identity_providers" , these check boxes(in the cognito aws console) should not get deselected.

### Actual Behavior

1. Modifies everytime
2. Selects and deselects the "allowed_oauth_flows, allowed_oauth_scopes, supported_identity_providers" options on alternative applys.

### Steps to Reproduce

1. Create a cognito user pool
2. Create an cognito client in the user pool with these options 
```
resource "aws_cognito_user_pool_client" "client" {
  name         = "name"
  count        = "1"
  user_pool_id = "${aws_cognito_user_pool.pool.id}" # the cognito pool id created in step 1

  generate_secret                      = true
  explicit_auth_flows                  = ["ADMIN_NO_SRP_AUTH", "USER_PASSWORD_AUTH"]
  allowed_oauth_flows_user_pool_client = true
  allowed_oauth_flows                  = ["code"]
  supported_identity_providers         = ["COGNITO"]
  allowed_oauth_scopes                 = ["aws.cognito.signin.user.admin", "openid", "profile", "email"]

  callback_urls = ["https://example.com"  ]
}
```

1. terraform apply - This will select the "Cognito" and "allowed oauth scopes" check boxes in the aws cognito console in the "app client settings" section.
2. terraform apply - Apply again and you will find those options deselected

**Comment by evan-10e** [comment]
This issue looks related to the way the boto api works on cognito app clients. see https://github.com/boto/boto3/issues/1607

It appears that **all attributes must be provided** on app client updates. The problem is that Terraform constructs a partial set of attributes that is given to the update. (see source: https://github.com/terraform-providers/terraform-provider-aws/blob/5fca4227f68d7bd9c9befc9a151469f38537b516/aws/resource_aws_cognito_user_pool_client.go#L265-L325)

Unprovided attributes are removed during the update call.

**Comment by karthikkp** [comment]
@evan-10e Any work arounds?

**Comment by evan-10e** [comment]
@karthikkp 

Yes.

See https://forums.aws.amazon.com/thread.jspa?threadID=285102

**Comment by scala-guy** [comment]
I am encountering the same issue. Whenever I update any value (e.g. Callback URL, Signout URL etc.), the unmodified attributes are unchecked.

**Comment by ravenscar** [comment]
That's not much of a workaround as you need to manually merge everything.

This can be replicated in terraform without changing anything in the AWS console. Do an initial apply then change something (e.g. the callback url) and apply again. Everything except the callback URL will be unset. If you then apply again everything will get set except the callback URL - and now you are stuck in a loop where terraform will never apply all the changes.

The workaround I have used is that if I am making any change in a `aws_cognito_user_pool_client` resource I clear everything first, then allow terraform to set everything.

You can do this via the cli:
```
aws cognito-idp update-user-pool-client --user-pool-id <user-pool-id> --client-id <client-id>
```

No values are set in the call, this clears all the values. You can now run `terraform apply` and it will set all the values and won't get stuck in a loop.

**Comment by bflad** [comment]
The fix for this has been merged into master and will release with version 1.32.0 of the AWS provider, likely middle of next week.

**Comment by ilyash-b** [comment]
Thanks, @bflad

**Comment by bflad** [comment]
This has been released in [version 1.32.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1320-august-16-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by karthikkp** [comment]
The same behavior is exhibited even for userpool.
